### PR TITLE
QL: Combine multiple equal queries into In

### LINKED ISSUE #46477: StackOverflowError caused by deeply nested queries
## Environment

**Elasticsearch Versions:**

- Version: 6.8.2, Build: default/docker/b506955/2019-07-24T15:24:41.545295Z, JVM: 12.0.1
- Version: 7.3.1, Build: default/docker/4749ba6/2019-08-19T20:19:25.651794Z, JVM: 12.0.2

**Plugins Installed**: None

**JVM Versions:**

- OpenJDK Version "12.0.1" 2019-04-16 (build 12.0.1+12, mixed mode, sharing)
- OpenJDK Version "12.0.2" 2019-07-16 (build 12.0.2+10, mixed mode, sharing)

**OS Version:** Linux 4.15.0-42

## Problem

**Summary:** With some deliberately constructed queries (e.g.: 1000-level nested bool queries), an Elasticsearch node could be crashed consistently by `StackOverflowError` (thrown from Lucene modules).

**Expected Behavior:** Do not crash. Respond with errors to clients instead.

**Steps to reproduce**:

1. start a single-node cluster with default settings in Docker container (listening on http://localhost:9200):

    <details>

    ```sh
    VERSION="${ES_VERSION:-7.3.1}"
    IMAGE="docker.elastic.co/elasticsearch/elasticsearch:${VERSION}"

    docker run --rm \
        --name es-bug \
        --net=host \
        -e 'discovery.type=single-node' \
        "${IMAGE}"
    ```

    </details>

1. create index with the following mappings:

    <details>

    ```json
    {
      "mappings": {
        "properties": {
          "item_id": {
            "type": "long"
          }
        }
      },
      "settings": {
        "index": {
          "number_of_replicas": "0",
          "number_of_shards": "5"
        }
      }
    }
    ```

    </details>

1. construct an SQL query with 1000 `OR` clauses (which looks like `SELECT * FROM sample WHERE (item_id=1 OR item_id=2 OR item_id=3 ... OR item_id=10) LIMIT 1000`), and translate it to JSON query (output at `/tmp/xxxx.json`):

    <details>

    ```sh
    #!/bin/bash

    function concat_clauses {
        local n="$1"

        for i in $(seq 1 "${n}")
        do
            echo "item_id=${i}"
        done | paste -s -d@ - - | sed 's/@/ OR /g'
    }

    function translate_query {
        local n="$1"
        local sql="$(printf 'SELECT * FROM sample WHERE (%s) LIMIT 1000' "$(concat_clauses "${n}")")"
        local output="$(mktemp --suffix=.json)"

        curl -Sfs -XPOST \
            -H 'Content-Type: application/json' \
            -d "$(printf '{"query": "%s"}' "${sql}")" \
            -o "${output}" \
            http://localhost:9200/_xpack/sql/translate

        echo "${output}"
    }

    translate_query 1000
    ```

    </details>

1. execute the query (replace `/tmp/xxxx.json` with actual path):

    <details>

    ```sh
    curl -Sfs -XPOST \
        -H 'Content-Type: application/json' \
        -d /tmp/xxxx.json \
        http://localhost:9200/sample/_search
    ```

    </details>

1. the cluster should crash with `StackOverflowError`.

**Logs:**

1. Logs of 6.8.2 cluster:

    <details>

    ```
    [2019-08-30T10:23:54,405][INFO ][o.e.e.NodeEnvironment    ] [2Hu963v] heap size [989.8mb], compressed ordinary object pointers [true]
    [2019-08-30T10:23:54,407][INFO ][o.e.n.Node               ] [2Hu963v] version[6.8.2], pid[1], build[default/docker/b506955/2019-07-24T15:24:41.545295Z], OS[Linux/4.15.0-42-generic/amd64], JVM[Oracle Corporation/OpenJDK 64-Bit Server VM/12.0.1/12.0.1+12]
    [2019-08-30T10:23:54,407][INFO ][o.e.n.Node               ] [2Hu963v] JVM arguments [-Xms1g, -Xmx1g, -XX:+UseConcMarkSweepGC, -XX:CMSInitiatingOccupancyFraction=75, -XX:+UseCMSInitiatingOccupancyOnly, -Des.networkaddress.cache.ttl=60, -Des.networkaddress.cache.negative.ttl=10, -XX:+AlwaysPreTouch, -Xss1m, -Djava.awt.headless=true, -Dfile.encoding=UTF-8, -Djna.nosys=true, -XX:-OmitStackTraceInFastThrow, -Dio.netty.noUnsafe=true, -Dio.netty.noKeySetOptimization=true, -Dio.netty.recycler.maxCapacityPerThread=0, -Dlog4j.shutdownHookEnabled=false, -Dlog4j2.disable.jmx=true, -Djava.io.tmpdir=/tmp/elasticsearch-477903446703128873, -XX:+HeapDumpOnOutOfMemoryError, -XX:HeapDumpPath=data, -XX:ErrorFile=logs/hs_err_pid%p.log, -Xlog:gc*,gc+age=trace,safepoint:file=logs/gc.log:utctime,pid,tags:filecount=32,filesize=64m, -Djava.locale.providers=COMPAT, -XX:UseAVX=2, -Des.cgroups.hierarchy.override=/, -Des.path.home=/usr/share/elasticsearch, -Des.path.conf=/usr/share/elasticsearch/config, -Des.distribution.flavor=default, -Des.distribution.type=docker]
    [2019-08-30T10:24:35,837][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [2Hu963v] fatal error in thread [elasticsearch[2Hu963v][search][T#1]], exiting
    java.lang.StackOverflowError: null
        at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:212) ~[lucene-core-7.7.0.jar:7.7.0 8c831daf4eb41153c25ddb152501ab5bae3ea3d5 - jimczi - 2019-02-04 23:16:28]
        at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:246) ~[lucene-core-7.7.0.jar:7.7.0 8c831daf4eb41153c25ddb152501ab5bae3ea3d5 - jimczi - 2019-02-04 23:16:28]
        at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:246) ~[lucene-core-7.7.0.jar:7.7.0 8c831daf4eb41153c25ddb152501ab5bae3ea3d5 - jimczi - 2019-02-04 23:16:28]
        at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:246) ~[lucene-core-7.7.0.jar:7.7.0 8c831daf4eb41153c25ddb152501ab5bae3ea3d5 - jimczi - 2019-02-04 23:16:28]
        at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:246) ~[lucene-core-7.7.0.jar:7.7.0 8c831daf4eb41153c25ddb152501ab5bae3ea3d5 - jimczi - 2019-02-04 23:16:28]
        at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:246) ~[lucene-core-7.7.0.jar:7.7.0 8c831daf4eb41153c25ddb152501ab5bae3ea3d5 - jimczi - 2019-02-04 23:16:28]
        at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:246) ~[lucene-core-7.7.0.jar:7.7.0 8c831daf4eb41153c25ddb152501ab5bae3ea3d5 - jimczi - 2019-02-04 23:16:28]
        at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:246) ~[lucene-core-7.7.0.jar:7.7.0 8c831daf4eb41153c25ddb152501ab5bae3ea3d5 - jimczi - 2019-02-04 23:16:28]
        at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:246) ~[lucene-core-7.7.0.jar:7.7.0 8c831daf4eb41153c25ddb152501ab5bae3ea3d5 - jimczi - 2019-02-04 23:16:28]

    (logs truncated because the same line is repeated 2000+ times)
    ```

    </details>

1. Logs of 7.3.1 cluster:

    <details>

    ```
    {"type": "server", "timestamp": "2019-08-30T10:25:27,743+0000", "level": "INFO", "component": "o.e.e.NodeEnvironment", "cluster.name": "docker-cluster", "node.name": "foo",  "message": "heap size [989.8mb], compressed ordinary object pointers [true]"  }
    {"type": "server", "timestamp": "2019-08-30T10:25:27,745+0000", "level": "INFO", "component": "o.e.n.Node", "cluster.name": "docker-cluster", "node.name": "foo",  "message": "version[7.3.1], pid[1], build[default/docker/4749ba6/2019-08-19T20:19:25.651794Z], OS[Linux/4.15.0-42-generic/amd64], JVM[Oracle Corporation/OpenJDK 64-Bit Server VM/12.0.2/12.0.2+10]"  }
    {"type": "server", "timestamp": "2019-08-30T10:25:27,746+0000", "level": "INFO", "component": "o.e.n.Node", "cluster.name": "docker-cluster", "node.name": "foo",  "message": "JVM arguments [-Xms1g, -Xmx1g, -XX:+UseConcMarkSweepGC, -XX:CMSInitiatingOccupancyFraction=75, -XX:+UseCMSInitiatingOccupancyOnly, -Des.networkaddress.cache.ttl=60, -Des.networkaddress.cache.negative.ttl=10, -XX:+AlwaysPreTouch, -Xss1m, -Djava.awt.headless=true, -Dfile.encoding=UTF-8, -Djna.nosys=true, -XX:-OmitStackTraceInFastThrow, -Dio.netty.noUnsafe=true, -Dio.netty.noKeySetOptimization=true, -Dio.netty.recycler.maxCapacityPerThread=0, -Dlog4j.shutdownHookEnabled=false, -Dlog4j2.disable.jmx=true, -Djava.io.tmpdir=/tmp/elasticsearch-14577330249979445676, -XX:+HeapDumpOnOutOfMemoryError, -XX:HeapDumpPath=data, -XX:ErrorFile=logs/hs_err_pid%p.log, -Xlog:gc*,gc+age=trace,safepoint:file=logs/gc.log:utctime,pid,tags:filecount=32,filesize=64m, -Djava.locale.providers=COMPAT, -Des.cgroups.hierarchy.override=/, -Dio.netty.allocator.type=unpooled, -XX:MaxDirectMemorySize=536870912, -Des.path.home=/usr/share/elasticsearch, -Des.path.conf=/usr/share/elasticsearch/config, -Des.distribution.flavor=default, -Des.distribution.type=docker, -Des.bundled_jdk=true]"  }
    {"type": "server", "timestamp": "2019-08-30T10:25:42,163+0000", "level": "ERROR", "component": "o.e.b.ElasticsearchUncaughtExceptionHandler", "cluster.name": "docker-cluster", "node.name": "foo", "cluster.uuid": "1QZlVdepSh6FnfX7CVGGNA", "node.id": "jh-YTT5IQ6itcnMbbyjUww",  "message": "fatal error in thread [elasticsearch[foo][search][T#2]], exiting" , 
    "stacktrace": ["java.lang.StackOverflowError: null",
    "at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:221) ~[lucene-core-8.1.0.jar:8.1.0 dbe5ed0b2f17677ca6c904ebae919363f2d36a0a - ishan - 2019-05-09 19:34:03]",
    "at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:255) ~[lucene-core-8.1.0.jar:8.1.0 dbe5ed0b2f17677ca6c904ebae919363f2d36a0a - ishan - 2019-05-09 19:34:03]",
    "at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:255) ~[lucene-core-8.1.0.jar:8.1.0 dbe5ed0b2f17677ca6c904ebae919363f2d36a0a - ishan - 2019-05-09 19:34:03]",
    "at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:255) ~[lucene-core-8.1.0.jar:8.1.0 dbe5ed0b2f17677ca6c904ebae919363f2d36a0a - ishan - 2019-05-09 19:34:03]",
    "at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:255) ~[lucene-core-8.1.0.jar:8.1.0 dbe5ed0b2f17677ca6c904ebae919363f2d36a0a - ishan - 2019-05-09 19:34:03]",
    "at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:255) ~[lucene-core-8.1.0.jar:8.1.0 dbe5ed0b2f17677ca6c904ebae919363f2d36a0a - ishan - 2019-05-09 19:34:03]",
    "at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:255) ~[lucene-core-8.1.0.jar:8.1.0 dbe5ed0b2f17677ca6c904ebae919363f2d36a0a - ishan - 2019-05-09 19:34:03]",
    "at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:255) ~[lucene-core-8.1.0.jar:8.1.0 dbe5ed0b2f17677ca6c904ebae919363f2d36a0a - ishan - 2019-05-09 19:34:03]",
    "at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:255) ~[lucene-core-8.1.0.jar:8.1.0 dbe5ed0b2f17677ca6c904ebae919363f2d36a0a - ishan - 2019-05-09 19:34:03]",

    (logs truncated because the same line is repeated 2000+ times)
    ```

    </details>

### LINKED ISSUE #62804: QL: Improve disjunction translations on the same field
Queries with a lot of `OR` clauses on the same field, end up being translated to individual bool clauses for every single expression. While correct this can result in very large, non-optimal requests which in turn can lead to other problems (see #46477).
QL can optimize this scenario away so the multiple bool queries are translated into a terms one,

### PR DESCRIPTION
**Pull request by costin**
Some queries can have an excessive number of OR clauses on the same
field (typically by being generated). This can lead to large query
requests that can cause SO errors.
This PR tries to address this pattern by combining the equals into an
In expression which in turn gets compressed into a terms query vs one
term query per entry.

Fix #62804
Fix #46477

### CODE REVIEW

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java [hunk 1]
```diff
@@ -1036,6 +1042,79 @@ private static boolean notEqualsIsRemovableFromConjunction(NotEquals notEquals,
...
+                            zoneId = in.zoneId();
+                        }
+                    } else {
+                        ors.add(exp);
+                    }
+                }
+
+                if (found.isEmpty() == false) {
+                    // combine equals alongside the existing ors
+                    ZoneId finalZoneId = zoneId;
```
Comment (astefan) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:1097]: Why do you need a second zoneid variable? You could use `zoneId`.

Comment (bpintea) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:1097]: nit: any reason to use the proxy `finalZoneId`? good practice?

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java [hunk 2]
```diff
@@ -106,6 +107,7 @@
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.startsWith;
 
+@TestLogging(value = "org.elasticsearch.xpack.sql:TRACE", reason = "debug")
```
Comment (astefan) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:110]: Leftover?

**Review by astefan — APPROVED** [review]
LGTM. Left two comments.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java [hunk 3]
```diff
@@ -1036,6 +1042,79 @@ private static boolean notEqualsIsRemovableFromConjunction(NotEquals notEquals,
...
+                List<Expression> exps = splitOr(e);
+
+                Map<Expression, Set<Expression>> found = new LinkedHashMap<>();
+                ZoneId zoneId = null;
+                List<Expression> ors = new ArrayList<>();
+
+                for (Expression exp : exps) {
+                    if (exp instanceof Equals) {
+                        Equals eq = (Equals) exp;
+                        // consider only equals against foldables
```
Comment (bpintea) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:1074]: QQ: I guess the `IN` could contain non foldable expressions? Wouldn't a similar check be useful for IN's elements?

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java (hunk 1 above)
Comment (costin) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:1097]: Lambda usage - the variable needs to be final...

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java (hunk 2 above)
Comment (bpintea) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:110]: was this left intentionally?

**Review by bpintea — APPROVED** [review]
lgtm, only left some prospecting questions.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java (hunk 1 above)
Comment (astefan) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:1097]: Doesn't hurt to have a `final` in front of it ;-).

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java (hunk 3 above)
Comment (costin) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:1075]: `In` requires all values to be foldable and checks this inside its `resolveType`. Hence why any `In` produced will only have foldable expressions.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java (hunk 1 above)
Comment (costin) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java]: See the previous comment

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java (hunk 2 above)
Comment (costin) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java]: See Andrei's comment

File: x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRulesTests.java [hunk 4]
```diff
@@ -1377,10 +1382,131 @@ public void testExactMatchRLike() throws Exception {
...
+
+        Or firstOr = new Or(EMPTY, new In(EMPTY, fa, singletonList(one)), lessThanOf(fa, two));
+        Or secondOr = new Or(EMPTY, firstOr, new In(EMPTY, fa, singletonList(three)));
+        Expression e = new CombineDisjunctionsToIn().rule(secondOr);
+        assertEquals(Or.class, e.getClass());
+        Or or = (Or) e;
+        assertEquals(LessThan.class, or.right().getClass());
+        LessThan lt = (LessThan) or.right();
+        assertEquals(fa, lt.left());
+        assertEquals(two, lt.right());
```
Comment (palesz) [inline x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRulesTests.java:1506]: Super nit-picky. Would just test that there are no changes:

```suggestion
        assertEquals(or.right(), firstOr.right());
```

**Review by palesz — CHANGES_REQUESTED** [review]
One typo in the tests that needs to be fixed, otherwise LGTM.

**Comment by costin** [comment]
The tests keep on failing due to #65419 - depending on the expression, either the rule gets triggered constantly or the unit test fails. I'll merge this after the other PR to account for the change.

**Comment by costin** [comment]
@elasticmachine update branch

**Pull request merged**